java/compaction/compaction-job-execution/src/main/java/sleeper/compaction/job/execution/CompactionTask.java

```diff
@@ -0,0 +1,123 @@
+    private final MessageReceiver messageReceiver;
+    private final MessageConsumer messageConsumer;
+    private final FailedJobHandler failedJobHandler;
+
+    public CompactionTask(InstanceProperties instanceProperties, Supplier<Instant> timeSupplier,
+            MessageReceiver messageReceiver, MessageConsumer messageConsumer, FailedJobHandler failedJobHandler) {
+        maxTimeInSeconds = instanceProperties.getInt(COMPACTION_TASK_MAX_TIME_IN_SECONDS);
+        maxConsecutiveFailures = instanceProperties.getInt(COMPACTION_TASK_MAX_CONSECUTIVE_FAILURES);
+        this.timeSupplier = timeSupplier;
+        this.messageReceiver = messageReceiver;
+        this.messageConsumer = messageConsumer;
+        this.failedJobHandler = failedJobHandler;
+    }
+
+    public void runAt(Instant startTime) throws InterruptedException, IOException {
+        Instant maxTime = startTime.plus(Duration.ofSeconds(maxTimeInSeconds));
+        int numConsecutiveFailures = 0;
+        Instant currentTime = startTime;
+        long totalNumberOfMessagesProcessed = 0;
+        while (numConsecutiveFailures < maxConsecutiveFailures && currentTime.isBefore(maxTime)) {
+            Optional<JobAndMessage> jobAndMessageOpt = messageReceiver.receiveMessage();
+            if (!jobAndMessageOpt.isPresent()) {
+                numConsecutiveFailures++;
+            } else {
+                JobAndMessage jobAndMessage = jobAndMessageOpt.get();
+                LOGGER.info("CompactionJob is: {}", jobAndMessage.job);
+                try {
+                    messageConsumer.consume(jobAndMessage);
+                    totalNumberOfMessagesProcessed++;
+                    numConsecutiveFailures = 0;
+                } catch (Exception e) {
+                    LOGGER.error("Failed processing compaction job, putting job back on queue", e);
+                    numConsecutiveFailures++;
+                    failedJobHandler.onFailure(jobAndMessage);
+                }
+            }
+            currentTime = timeSupplier.get();
+        }
+        if (numConsecutiveFailures >= maxConsecutiveFailures) {
+            LOGGER.info("Terminating compaction task as {} consecutive failures exceeds maximum of {}",
+                    numConsecutiveFailures, maxConsecutiveFailures);
+        } else {
+            LOGGER.info("Terminating compaction task as it ran for {}, exceeeding maximum of {}",
+                    LoggedDuration.withFullOutput(startTime, currentTime),
+                    LoggedDuration.withFullOutput(Duration.ofSeconds(maxTimeInSeconds)));
+        }
+        LOGGER.info("Total number of messages processed = {}", totalNumberOfMessagesProcessed);
+    }
+
+    @FunctionalInterface
+    interface MessageReceiver {
+        Optional<JobAndMessage> receiveMessage() throws InterruptedException, IOException;
+    }
+
+    @FunctionalInterface
+    interface MessageConsumer {
+        void consume(JobAndMessage jobAndMessage) throws Exception;
```

Comment:
It should be possible to set this up so that we don't need to pass in the Message object here. The SQS code should be able to handle that.